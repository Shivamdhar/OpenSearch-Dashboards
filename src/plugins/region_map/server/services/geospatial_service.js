/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export default class GeospatialService {
  constructor(esDriver) {
    this.esDriver = esDriver;
  }

  uploadGeojson = async (context, req, res) => {
    try {
      const { callAsCurrentUser } = await this.esDriver.asScoped(req);
      const uploadResponse = await callAsCurrentUser('geo.geoQuery', { body: req.body });
      return res.ok({
        body: {
          ok: true,
          resp: uploadResponse,
        },
      });
    } catch (err) {
      return res.ok({
        body: {
          ok: false,
          resp: err.message,
        },
      });
    }
  };
}
