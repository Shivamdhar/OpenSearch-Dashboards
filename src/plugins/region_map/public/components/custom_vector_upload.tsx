/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useState, Fragment, useEffect } from 'react';
import { EuiButton } from '@elastic/eui';
import { Services, getServices } from '../services';
import EmptyPrompt from './empty_prompt';
import VectorUploadOptions from './vector_upload_options';

export type CustomVectorUploadProps = {
  getServiceSettings: () => Promise<IServiceSettings>;
} & VisOptionsProps<RegionMapVisParams>;

function CustomVectorUpload(props: CustomVectorUploadProps) {
  const services = getServices(props.vis.http);
  const GEOSPATIAL_PLUGIN = 'opensearch-geospatial';
  const [isGeospatialPluginInstalled, setIsGespatialPluginInstalled] = React.useState(false);

  // Logic for implementing empty state pattern
  // In case geospatial plugin is not installed, an empty prompt with details is shown.
  useEffect(() => {
    const checkIfGeospatialPluginIsInstalled = async () => {
      const result = await services.getPlugins();
      const installedPlugins = result.resp.map((plugin) => {
        return plugin.component;
      });
      if (installedPlugins.includes(GEOSPATIAL_PLUGIN)) {
        return true;
      }

      return false;
    };

    async function getPluginInfo() {
      const installedFlag = await Promise.resolve(checkIfGeospatialPluginIsInstalled());
      if (installedFlag) {
        setIsGespatialPluginInstalled(true);
      } else {
        setIsGespatialPluginInstalled(false);
      }
    }
    getPluginInfo();
  }, [services]);

  const emptyPromptProps = {
    iconType: 'popout',
    title: 'Missing geospatial plugin',
    bodyFragment: 'Install the geospatial plugin in order to upload custom vector maps.',
    actions: (
      <EuiButton href="https://opensearch.org/docs/latest/" color="primary" fill>
        Learn more
      </EuiButton>
    ),
  };

  return (
    <div>
      {isGeospatialPluginInstalled ? (
        <VectorUploadOptions {...props} />
      ) : (
        <EmptyPrompt {...emptyPromptProps} />
      )}
    </div>
  );
}

export { CustomVectorUpload };
