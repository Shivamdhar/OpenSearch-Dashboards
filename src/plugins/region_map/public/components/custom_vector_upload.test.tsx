/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import renderer, { act, create } from 'react-test-renderer';
import * as React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { CustomVectorUpload } from './custom_vector_upload';
import * as serviceApiCalls from '../services';

describe('custom vector upload component', () => {
  it('renders the empty prompt if geospatial plugin is installed', async () => {
    const mockDataWhenPluginIsUninstalled = {
      ok: true,
      resp: [
        {
          name: 'integTest-0',
          component: 'opensearch',
          version: '2.1.0.0-SNAPSHOT',
        },
      ],
    };
    const props = {
      vis: {
        http: '',
        notifications: '',
      },
    };
    jest.spyOn(serviceApiCalls, 'getServices').mockImplementation(() => {
      return {
        getPlugins: () => {
          return mockDataWhenPluginIsUninstalled;
        },
      };
    });
    let tree;
    await act(async () => {
      tree = renderer.create(<CustomVectorUpload {...props} />);
    });
    expect(tree.toJSON().children[0].props.testId).toBe('empty-prompt-id');
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
