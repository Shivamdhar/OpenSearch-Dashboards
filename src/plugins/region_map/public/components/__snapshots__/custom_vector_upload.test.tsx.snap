// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom vector upload component renders the empty prompt if geospatial plugin is installed 1`] = `
<div>
  <div
    className="euiEmptyPrompt"
    testId="empty-prompt-id"
  >
    <span
      color="subdued"
      data-euiicon-type="popout"
      size="xxl"
    />
    <div
      className="euiSpacer euiSpacer--m"
    />
    <h2
      className="euiTitle euiTitle--medium"
    >
      Missing geospatial plugin
    </h2>
    <span
      className="euiTextColor euiTextColor--subdued"
    >
      <div
        className="euiSpacer euiSpacer--m"
      />
      <div
        className="euiText euiText--medium"
      >
        <p>
          Install the geospatial plugin in order to upload custom vector maps.
        </p>
      </div>
    </span>
    <div
      className="euiSpacer euiSpacer--l"
    />
    <a
      className="euiButton euiButton--primary euiButton--fill"
      disabled={false}
      href="https://opensearch.org/docs/latest/"
      rel="noreferrer"
      style={
        Object {
          "minWidth": undefined,
        }
      }
    >
      <span
        className="euiButtonContent euiButton__content"
      >
        <span
          className="euiButton__text"
        >
          Learn more
        </span>
      </span>
    </a>
  </div>
</div>
`;
